File: backend/src/main/java/com/prueba/backend/service/EmployeeService.java
package com.prueba.backend.service;

import com.prueba.backend.dto.RequestDTO.EmployeeRequestDTO;
import com.prueba.backend.dto.ResponseDTO.EmployeeResponseDTO;

import java.util.List;

/**
 * Contrato de servicios para la gestión de empleados.
 *
 * <p>Define las operaciones CRUD disponibles para la entidad {@code Employee}.
 * La implementación concreta se encuentra en {@link com.prueba.backend.service.impl.EmployeeServiceImpl}.
 */
public interface EmployeeService {

    /**
     * Obtiene todos los empleados registrados en el sistema.
     *
     * @return lista de {@link EmployeeResponseDTO} con todos los empleados
     */
    List<EmployeeResponseDTO> findAll();

    /**
     * Busca un empleado por su identificador único.
     *
     * @param id identificador del empleado
     * @return {@link EmployeeResponseDTO} con los datos del empleado encontrado
     * @throws com.prueba.backend.exception.ResourceNotFoundException si el empleado no existe
     */
    EmployeeResponseDTO findById(Long id);

    /**
     * Crea y persiste un nuevo empleado. También obtiene las coordenadas
     * geográficas de la ciudad indicada mediante la API de Nominatim.
     *
     * @param dto datos de entrada del nuevo empleado
     * @return {@link EmployeeResponseDTO} con los datos del empleado creado
     * @throws com.prueba.backend.exception.DuplicateResourceException si el correo ya está registrado
     */
    EmployeeResponseDTO save(EmployeeRequestDTO dto);

    /**
     * Actualiza los datos de un empleado existente.
     * Si la ciudad cambia, se consultan nuevamente las coordenadas geográficas.
     *
     * @param id  identificador del empleado a actualizar
     * @param dto nuevos datos del empleado
     * @return {@link EmployeeResponseDTO} con los datos actualizados
     * @throws com.prueba.backend.exception.ResourceNotFoundException  si el empleado no existe
     * @throws com.prueba.backend.exception.DuplicateResourceException si el nuevo correo ya está en uso
     */
    EmployeeResponseDTO update(Long id, EmployeeRequestDTO dto);

    /**
     * Elimina un empleado y todos sus beneficios asociados (cascada).
     *
     * @param id identificador del empleado a eliminar
     * @throws com.prueba.backend.exception.ResourceNotFoundException si el empleado no existe
     */
    void deleteById(Long id);
}

File: frontend/src/services/employeeService.js
import api from "../api/axios";

/**
 * Servicio de acceso a la API REST para la gestión de empleados.
 * Encapsula todas las peticiones HTTP relacionadas con el recurso `/employees`.
 */
export default {
  /**
   * Obtiene la lista completa de empleados.
   * @returns {Promise} respuesta de Axios con `data` como arreglo de empleados
   */
  getEmployees() {
    return api.get("/employees");
  },

  /**
   * Obtiene los datos de un empleado específico.
   * @param {number|string} id - Identificador del empleado
   * @returns {Promise} respuesta de Axios con `data` como objeto empleado
   */
  getEmployeeById(id) {
    return api.get(`/employees/${id}`);
  },

  /**
   * Crea un nuevo empleado.
   * @param {Object} employee - Datos del empleado según el esquema del backend
   * @returns {Promise} respuesta de Axios con el empleado creado en `data`
   */
  createEmployee(employee) {
    return api.post("/employees", employee);
  },

  /**
   * Actualiza un empleado existente.
   * @param {number|string} id - Identificador del empleado a actualizar
   * @param {Object} employee - Nuevos datos del empleado
   * @returns {Promise} respuesta de Axios con el empleado actualizado en `data`
   */
  updateEmployee(id, employee) {
    return api.put(`/employees/${id}`, employee);
  },

  /**
   * Elimina un empleado por su ID.
   * @param {number|string} id - Identificador del empleado a eliminar
   * @returns {Promise} respuesta de Axios (204 No Content)
   */
  deleteEmployee(id) {
    return api.delete(`/employees/${id}`);
  },
};
